geode-core/src/main/java/org/apache/geode/distributed/internal/membership/gms/GMSUtil.java

```diff
@@ -51,6 +64,35 @@
     return parseLocators(locatorsString, addr);
   }
 
+  public static GMSMember readMemberID(DataInput in) throws IOException, ClassNotFoundException {
+    Object id = DataSerializer.readObject(in);
+    if (id == null || id instanceof GMSMember) {
+      return (GMSMember) id;
+    }
+    // return ((GMSMemberAdapter)((InternalDistributedMember)id).getNetMember()).getGmsMember();
+    try {
+      Method getNetMember = id.getClass().getMethod("getNetMember");
```

Comment:
Why is this method using reflection to invoke methods?

Reply:
It's using reflection because gms classes can't reference InternalDistributedMember or its getNetMember method.  This is backward-compatibility code for rolling-upgrade support.

Reply:
wrt GMSMembershipManager being an adapter, it doesn't actually make any GMS decisions - doesn't elect a coordinator doesn't decide whether a member is there or not, etc.  It mostly handles "surprise members" and "shunned members" and the TCPConduit communications service.  It needs to know about HighPriorityDistributionMessage, SerialDistributionMessages and the like.
I put a lot of effort into trying to make it not be an adapter and not know about InternalDistributedMember before deciding it wasn't an integral part of the gms package and that the Manager interface in gms is a contract between gms and the user of gms (Geode's DistributionManager).
This is completely in line with the old GemFire source base that used JGroups for gms.  In that implementation JGroupMembershipManager was the glue between GemFire and JGroups and translated between JGroups views/identifiers/messages and the versions used by GemFire.

Reply:
We're going to look into, in a separate effort, whether GMSMembershipManager can be moved into gms so that MembershipManager is the contract between Geode and membership.

Reply:
Be careful here. I know you say this is for backwards compatibility but invoking `getMethod` involves some allocations. If this or similar access is in a hot path consider caching the Method `object`.

Reply:
Good thing to consider Jake.   This method is only used for membership messages during rolling upgrade.  It isn't used for messages defined outside of the gms packages, like Geode cache operation messages.  After rolling upgrade it won't take the Method lookup path because gms messages use GMSMember identifiers instead of InternalDistributedMember identifiers.